[WIP] Use correct stresses in visco-elasto-plastic rheology

anne-glerum opened the pull request:
This PR is at the moment for discussing with everybody interested the changes to the elasticy and visco-plastic plugins that @bobmyhill, @naliboff and I discussed yesterday.  @dansand, your input would be great to have! 

FYI @EstherHeck 

### For all pull requests:

* [ ] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [ ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -252,11 +252,33 @@ namespace aspect
+                // every time the Stokes system is assembled.
+                //
+                // Moresi et al. (2003) use the full stresses at $t$ (Eq.  30) in the force term.
+                // There is also an incorrect minus sign in the force term.
```
Is this incorrect minus sign in the Moresi paper?

bobmyhill replied:
@naliboff in the Moresi paper, the force term is on the LHS (equations 2, 31). In ASPECT, the force term is on the RHS. We both use the same sign (i.e. - eta_eff/eta_el * [some tau]). So if we are interpreting the stress terms correctly, there must be a sign error.

bobmyhill replied:
We also need to use $tau^{t+\Delta te}$ to calculate the effective viscosity on the yield surface *for a non-damped elastic element* (l244 of rheology/visco_plastic.cc, eq. 36 of Moresi et al. 2003). This is also obtained from in.composition[i,j] (l114 of rheology/visco_plastic.cc).

bobmyhill replied:
On a related note, I think there should only be one elastic element at any evaluation point, rather than one-per-compositional-field-per-element. In other words, we should drag the elasticity calculations outside the loop over compositional fields and use a volume-averaged shear modulus. There isn't an obvious choice for whether we use harmonic (Reuss, isostress) or arithmetic (Voigt, isostrain) averaging for elasticity. I have a slight preference for arithmetic (probably better for small-strains).

anne-glerum replied:
> We also need to use $tau^{t+\Delta te}$ (rotated, advected) to calculate the effective viscosity on the yield surface _for a non-damped elastic element_ (l244 of rheology/visco_plastic.cc, eq. 36 of Moresi et al. 2003). This is also obtained from in.composition[i,j] (l114 of rheology/visco_plastic.cc).

Just to be sure, the stress in in.composition is only $tau^{t+\Delta te}$ when $\Delta te$ = $\Delta t$, right? Otherwise it would be $tau^{t+\Delta t}$.

anne-glerum replied:
When we evaluate the material model to compute the rhs force term, `in.composition` has been updated using the reaction_outputs. 

At the moment, Advection iterations are not allowed because this could indeed lead to a different `in.composition` in each iteration.

anne-glerum replied:
The force term could still be different during every non-linear iteration, because the velocity that goes into the Advection equation is updated. Already for single Advection schemes, the viscosity will be different in every Stokes iteration.

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -252,11 +252,33 @@ namespace aspect
+                // The force term should be computed as:
+                // $\frac{-\eta_{effcreep}  tau_0}{\eta_{e}}$, where $\eta_{effcreep}$ is the
+                // harmonic average of the vp and e viscosity.
+                // How will we get to this viscosity? 
```
That is even trickier if the viscosity we need to use is not just old the VEP or VE viscosity, but rather the harmonic average of all the different deformation mechanism viscosities. Actually, I think this should only be done if we need to locally iterate between the plastic and elastic stresses. If we use the standard viscosity rescaling method, then the old VEP viscosity should be correct.

To get the old viscosity, as discussed the options are:
1. Reevalaute the material model using the "old" values for the material model inputs
2. Store and keep track of the old viscosity in a compositional field.

In principal, I am fine with either solution.

bobmyhill replied:
These are the viscosities given the strain rate at t+delta t_e. So the question for me is how we obtain the strain rate at this time, given that we only have the strain rate at t+delta t. Do we just assume that the strain rate is constant between t and t + delta t_e? This would be inconsistent with both the time-stepping, and with the assumption that the spin and rate of change of stress are both constant between t and t + delta t_e.

bobmyhill replied:
We could extrapolate the strain rate linearly, which would imply that viscosities are independent of strain rate at each time step. That would be inconsistent with the rheological laws though.

bobmyhill replied:
I'm tempted to suggest that we concentrate on fixing the Delta t_e = Delta t case, and add a warning for people attempting to vary the elastic timestep.

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -252,11 +252,33 @@ namespace aspect
+                // based on the current_linearization_point. This means that it will be updated after every
+                // nonlinear Stokes iteration, and is ahead of the stresses that are used in the force term.
+                // Also, out.viscosities contains the vep viscosity, which is either the harmonic average
+                // of the viscous and elastic viscosity, or the plastic viscosity, not $\eta_{effcreep}$.
```
As noted above, until we start using a local iterative method, I think the viscosity value from the previous time step should be correct?

bobmyhill replied:
@naliboff If we use the old viscosities from the previous time step, then we could have elements that are spuriously yielding just because they yielded in the previous time step. 

> the harmonic average of the viscous and elastic viscosity, or the plastic viscosity, not $\eta_{effcreep}$.
@anne-glerum if the plastic damper is not being used, these are the same, aren't they? If the material is yielding,  then eta_effcreep is the plastic viscosity. If it isn't yielding, then eta_effcreep is the harmonic average of the viscous and elastic viscosity.

anne-glerum replied:
Yesterday, eta_effcreep in the Appendix seemed to be something else, but I see you've updated it now. So yes, out.viscosities returns either the harmonic average of the viscous and elastic viscosity, or the plastic viscosity, which is what is needed for eta_effcreep.

bobmyhill replied:
Yes, sorry about that... I copy-pasted from the previous section and didn't change everything I should have changed. I think the Appendices should be error free now <crosses fingers>.

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -301,6 +323,13 @@ namespace aspect
               {
                 // Get old stresses from compositional fields
+                //
+                // TODO: Also use the stresses from the old_solution, not just the strain rates.
```
👍

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -309,8 +338,8 @@ namespace aspect
-                // Average viscoelastic viscosity
-                const double average_viscoelastic_viscosity = out.viscosities[i];
+                // Average viscoelastoplastic viscosity
+                const double average_viscoelastoplastic_viscosity = out.viscosities[i];
```
👍

naliboff commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -240,6 +240,12 @@ namespace aspect
+                    // the ve stresses stored in in.composition are the rotated, averaged,
+                    // and advected stresses from the advection solve ($\tau^{t+\Delta t}$).
+                    // They should be the unaveraged, but advected stresses? How could we get
+                    // to those?
```
Hmmm, I can't think of any easy to get the unaveraged, but advected values. However, I'm still not clear why we need to use the unaveraged values here?

bobmyhill replied:
If we're doing anything at $\tau^{t+\Delta te}$ where delta te > delta t don't we have to advect into the future? It isn't clear to me how we should deal with Delta t != Delta t_e in ASPECT.

naliboff reviewed (commented):
@anne-glerum - Thanks for starting this discussion! I've added comments under most of your discussion points, but will wait to add further comments until @bobmyhill and @dansand have a chance to chime in.

bobmyhill commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -301,6 +323,13 @@ namespace aspect
+                // Explanantion: When this function is called for the assembly of the advection
+                // equations for the stress components, in.composition contains the
+                // values of the current_linearization_point, which is extrapolated
+                // from the old_solution and the old_old_solution.
```
Can you remind us what the old_solution and old_old_solution are?

anne-glerum replied:
These are the solutions for the governing variables obtained in the last and second to last timestep.

bobmyhill replied:
Thank you :)

bobmyhill commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -301,6 +323,13 @@ namespace aspect
                 // Get old stresses from compositional fields
+                //
+                // TODO: Also use the stresses from the old_solution, not just the strain rates.
+                // Explanantion: When this function is called for the assembly of the advection
```
```suggestion
                // Explanation: When this function is called for the assembly of the advection
```

bobmyhill commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -315,6 +321,14 @@ namespace aspect
+            // likely to be constant within a timestep, as through the nonlinear iterations the viscous
+            // viscosity (strain rate and pressure dependency) and the yield stress (pressure dependency)
+            // are likely to be updated. In this case, local iterations between the viscoelastic stress
+            // and the plastic stress need to be performed to find lambda.
```
I think I misspoke over zoom. As long as none of these variables are strain-dependent, it shouldn't be necessary to do any local iterations - the nonlinear iterations should already deal with pressure dependency, for example. 

The main issue is that the code does allow for strain dependent viscosity (i.e. the dislocation creep and Peierls options), and in these cases we need to calculate the stress state of the system, and calculate the viscosities using that stress.

anne-glerum replied:
Okay, I'll update that. But that means we should implement the local iterations.

naliboff commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -244,8 +248,6 @@ namespace aspect
-                    // They should be the unaveraged, but advected stresses? How could we get
-                    // to those?
                     const double viscoelastic_strain_rate_invariant = elastic_rheology.calculate_viscoelastic_strain_rate(in.strain_rate[i],
                                                                       stress_old,
```
@anne-glerum @bobmyhill - I wonder if here we should not be tau^t, but rather tau^(t+te)?

At present, we are using the updated strain rate, but the old viscoelastic stress. Maybe this could explain why the non-linear solvers are having so much trouble for problems with rapidly changing deformation fields?

naliboff replied:
To follow, this would propagate into:
```
current_edot_ii = std::max(viscoelastic_strain_rate_invariant,
                                               min_strain_rate);
```

and then 
`double current_stress = 2. * viscosity_pre_yield * current_edot_ii;`

anne-glerum replied:
When we need the viscosity for the Stokes assembly, in.composition holds tau^(t+dt), because it's after the Advection solve.

naliboff reviewed (commented):
More thoughts about where we might need to use different stress values.

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -146,8 +147,27 @@ namespace aspect
+        // If a damper is used, the current implementation in visco_plastic
+        // is not correct. The way the effective viscosity is computed is only valid
+        // in the absence of a damper and for a yield stress that can be expressed as
+        // some value of the second invariant of the deviatoric stress.
```
Is this solved by simply taking the harmonic average of the damper and viscoelastic viscosity?

anne-glerum replied:
No, at the moment we don't know how the damper could be correctly included.

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -146,8 +147,27 @@ namespace aspect
         if (this->convert_output_to_years())
           fixed_elastic_time_step *= year_in_seconds;
 
+        // When using the visco_plastic material model,
```
I think this issue also applies to the viscoelastic material as well?

anne-glerum replied:
The viscoelastic material model will also have to be updated due to changes in elasticity.cc, but I haven't done that yet. First want to be convinced vp works!

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -181,26 +201,17 @@ namespace aspect
+                     Parameters<dim>::NonlinearSolver::iterated_Advection_and_Newton_Stokes ||
                      this->get_parameters().nonlinear_solver ==
-                     Parameters<dim>::NonlinearSolver::single_Advection_iterated_defect_correction_Stokes),
+                     Parameters<dim>::NonlinearSolver::iterated_Advection_and_defect_correction_Stokes),
```
😄

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -228,6 +239,16 @@ namespace aspect
               std::make_unique<ElasticAdditionalOutputs<dim>> (n_points));
           }
+
+        // Create the ReactionRateOutputs that are necessary for the operator splitting
```
Should we enforce that operator splitting is used? Or maybe this enforced somewhere below?

anne-glerum replied:
Yep, I've added that.

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -246,22 +267,44 @@ namespace aspect
+                // contains the rotated and advected stresses $tau^{0adv}$.
+                // Only at the beginning of the next timestep do we add the stress update of the
+                // current timestep to the stress stored in the compositional fields, giving
+                // $\tau{t+\Delta te}$ (with $t+\Delta te}$ being the current timestep.
```
```suggestion
                // $\tau{t+\Delta te}$ with ${t+\Delta te}$ being the current timestep.
```

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -273,78 +316,168 @@ namespace aspect
                                               MaterialModel::MaterialModelOutputs<dim> &out) const
       {
-        if (in.current_cell.state() == IteratorState::valid && this->get_timestep_number() > 0 && in.requests_property(MaterialProperties::reaction_terms))
+        // TODO: also evaluate at t = 0?
```
I initially thought we would want to do this at t=0, but now I'm not sure it makes sense to do so if we have not actually computed a numerical time step yet? However, I think we would still definitely want to add the force term to the RHS of the momentum equation at t=0.

anne-glerum replied:
At the moment, we use the fixed_elastic_timestep when no numerical timestep is available yet, so that's probably fine. If the initial fields are zero, the reaction terms will be zero. Thinking about it again, I think the reaction terms should be computed at t0 just in case the initial stresses on the fields are not zero.

The force terms are already added at t0.

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -273,78 +316,168 @@ namespace aspect
+                // old stress have been incorrectly switched. In Eq. (8) of the same supplement,
+                // the advected stress history tensor is used, but this tensor is only advected during
+                // the advection solve for which we are computing the reaction terms here (in case of fields).
+                // TODO Change: delete these lines. Move to somewhere else?
```
I think these lines should be removed from the code, but the key points added into the ASPECT manual when we put in a section for the VEP rheology.

gassmoeller replied:
Can you explain again why these lines are not needed anymore? If they are unnecessary or wrong, they should be deleted (no need to keep them). If they are a feature that is unsupported with the new way we do elastic stressses, we could keep a note about this (which feature, which PR it was removed), but still remove the code to avoid outdated code and confusion (we can always get it back from the git history).

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -273,78 +316,168 @@ namespace aspect
+      // splitting step that at the beginning of the new timestep updates the
+      // stored compositions $tau^{0\mathrm{adv}}$ at time $t$ to $tau^{t}$.
+      // Make sure that the old_solution is used as input to the material model and that
+      // the reaction_rates are take into account the correct timestep that they will be
```
```suggestion
      // the reaction_rates take into account the correct timestep that they will be
```

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -273,78 +316,168 @@ namespace aspect
+      // $\tau^{t}_T = 2 \eta^{t}_effcreep (\dot{\varepsilon}^{t}_T + \frac{\tau^{t}_{0adv}}{2 G \Delta te}$.
+      // Compute difference $\tau^{t}_T - \tau^{t}_{0adv}$.
+      // Multiply with previous timestep, divide by current timestep.
+      // Can we Assert here that $tau^{t+\Delta te} == \eta_{eff} D^{t+\Delta te}_{eff}$?
```
I don't understand what the Assert here would be checking. Can you elaborate?

naliboff reviewed (commented):
@anne-glerum - Partial review at the end of the today. Bravo for getting this all implemented and particularly the comments, which very succinctly and clearly outlines the "old versus new" issue in the context of ASPECTs solvers (+ how things are being updated)! 👏 Independent of elasticity, the comments are an excellent resource for anyone trying to understand the solver schemes and solution updates!

anne-glerum commented:
Maybe @bobmyhill , @jdannberg or @gassmoeller can also give this a look?

naliboff commented on include/aspect/postprocess/stress_component_statistics.h:
```diff
@@ -0,0 +1,53 @@
+/*
+  Copyright (C) 2011 - 2019 by the authors of the ASPECT code.
```
```suggestion
  Copyright (C) 2011 - 2022 by the authors of the ASPECT code.
```

naliboff commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -298,61 +354,166 @@ namespace aspect
+                // current timestep.
+                //const double previous_dt = this->get_old_timestep();
+                const double current_dt  = this->get_timestep();
+                //const double timestep_compensation_factor = previous_dt / current_dt;
```
Can tis line, line 506 (`//const double previous_dt = this->get_old_timestep();`) be removed since they are longer used? Likewise, should the documentation above be updated to just include lines 501-505?

naliboff reviewed (commented):
@anne-glerum - I added a few more suggestions and questions to the PR, nearly entirely just regarding documentation. Overall, the PR is in great shape and really well documented (thank you!). I did not get through the full PR, but did review the key updates to elasticity.cc, etc.

A few additional general questions:
1. Can the stress_component_statistics still be removed, per a comment in the .cc file? If it does need to be included, I will review that as well.
2. Are the updates to stress.cc implementing the same changes as proposed in #4375?

anne-glerum commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -298,61 +354,166 @@ namespace aspect
-            for (unsigned int i=0; i < in.n_evaluation_points(); ++i)
+            // Only create the evaluator the first time we get here
+            // We assume (and assert in parse_parameters) that the n_independent_components
+            // tensor components are the first fields in the correct order.
```
I've kept this comment here, because we start the `evaluator_composition` with the first field (`this->introspection().component_indices.compositional_fields[0]`).

anne-glerum commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -276,18 +327,23 @@ namespace aspect
-            // Get old (previous time step) velocity gradients
+            // Get velocity gradients of the current timestep $t+dt=t+dte$
+            // and the compositions from the previous timestep $t$.
+            // Assume velocity and composition use the same quadrature.
```
Good point!

anne-glerum commented:
> @anne-glerum - I added a few more suggestions and questions to the PR, nearly entirely just regarding documentation. Overall, the PR is in great shape and really well documented (thank you!). I did not get through the full PR, but did review the key updates to elasticity.cc, etc.
> 
> A few additional general questions:
> 
>     1. Can the stress_component_statistics still be removed, per a comment in the .cc file? If it does need to be included, I will review that as well.
> 
>     2. Are the updates to stress.cc implementing the same changes as proposed in #4375?

No. The changes in #4375 are valid for the VEP implementation in main, the changes in stress.cc are for the VEP implementation in this PR.

gassmoeller commented on source/particle/property/elastic_stress.cc:
```diff
@@ -121,7 +121,7 @@ namespace aspect
 
         for (unsigned int i = 0; i < SymmetricTensor<2,dim>::n_independent_components ; ++i)
-          particle->get_properties()[data_position + i] += material_outputs.reaction_terms[0][i];
+          particle->get_properties()[data_position + i] = material_inputs.composition[0][i] + material_outputs.reaction_terms[0][i];
```
This is potentially a problem. By first interpolating the particle property to the field and then assigning back to the particle we introduce a lot of numerical diffusion. Why does the approach above not work? Because of the nonlinear solver?

anne-glerum replied:
I think my reasoning was as follows:
* In the operator splitting step, we update the solution (the stress compositional fields) of the previous timestep.
* After that we advance the particles and evaluate the updated solution at their location
* We then update the particle properties, in this case the stress. If we use the stress values on the particles, then these do not include the operator splitting update.

gassmoeller reviewed (commented):
I will need to finish this later, but I have a number of small items and questions at this stage.

gassmoeller commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -404,16 +619,31 @@ namespace aspect
+          dtc = std::min(std::min(this->get_parameters().maximum_time_step, this->get_parameters().maximum_first_time_step), elastic_timestep());
+        const double timestep_ratio = dtc / elastic_timestep();
+        const double elastic_viscosity = timestep_ratio * calculate_elastic_viscosity(shear_modulus);
+        const double creep_viscosity = viscosity_pre_yield; //timestep_ratio *  calculate_viscoelastic_viscosity(viscosity_pre_yield, shear_modulus);
```
What about this commented code? If not necessary please remove. If this is some sort of explanation, please move to a separate line and expand the explanation.

anne-glerum replied:
Yep, should be deleted.

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -281,12 +314,12 @@ namespace aspect
             // Step 5: plastic yielding
 
             // Determine if the pressure used in Drucker Prager plasticity will be capped at 0 (default).
-            // This may be necessary in models without gravity and when the dynamic stresses are much higher
+            // This may be necessary in models without gravity and the dynamic stresses are much higher
```
I think this `when` is correct, is it not?

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -313,6 +346,7 @@ namespace aspect
                   // rescale the viscosity back to yield surface
                   if (current_stress >= yield_stress)
                     {
+                      // TODO: Appendix E omits the sqrt of the viscoelastic_strain_rate_invariant.
```
This comment is not  helpful, because it is unclear what Appendix E refers to. Can you clarify the comment and why it is here?

anne-glerum replied:
Sorry, that's a reminder that I need to discuss it with Bob. I'll remove it.

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -330,6 +364,14 @@ namespace aspect
+            // proposed stress is zero (up to a certain tolerance).
+            // Explanation: The variables in Eq. 36 of Moresi et al. (2003) for the effective viscosity
+            // can depend on the strain rate. In this case (i.e. for dislocation creep or strain-dependent weakening),
+            // local iterations are needed to find to be performed to find lambda.
```
```suggestion
            // local iterations are needed to find lambda.
```

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -346,7 +388,12 @@ namespace aspect
                                                              );
             output_parameters.composition_viscosities[j] = std::min(std::max(viscosity_yield, minimum_viscosity_for_composition), maximum_viscosity_for_composition);
+
+            // TODO: Assert that:
```
Do we need to solve this TODO? If you want to assert that, you want to include a tolerance, e.g. `Assert (std::abs(sum - part_1 - part_2) < std::numeric_limits<double>::epsilon() * 100 * sum, ExcMessage(...))` or whatever tolerance is reasonable.

gassmoeller reviewed (commented):
A few more comments. Let me know if I should take a look at the postprocessors as well.

anne-glerum commented on benchmarks/viscoelastic_relaxation/viscoelastic_relaxation.prm:
```diff
@@ -0,0 +1,185 @@
+subsection Discretization
+  set Composition polynomial degree     = 2
+  set Stokes velocity polynomial degree = 2
+  set Temperature polynomial degree     = 1
```
Indeed, temperature is irrelevant. However, I've now set the degree back to 2, because when using only 1 cell (so no repetitions, no refinement), the temperature is not set correctly, giving an error of 0 K in the material model, even though temperature is set to 293 K. 
```
An error occurred in line <165> of file </home/bbpanneg/software/aspect/aspect_fix_stresses_elasticity/aspect/source/material_model/rheology/visco_plastic.cc> in function
    aspect::MaterialModel::IsostrainViscosities aspect::MaterialModel::Rheology::ViscoPlastic<dim>::calculate_isostrain_viscosities(const aspect::MaterialModel::MaterialModelInputs<dim>&, unsigned int, const std::vector<double, std::allocator<double> >&, const std::vector<double, std::allocator<double> >&, const std::vector<unsigned int>&) const [with int dim = 2]
The violated condition was: 
    temperature_for_viscosity != 0
Additional information: 
    The temperature used in the calculation of the visco-plastic rheology
    is zero. This is not allowed, because this value is used to divide
    through. It is probably being caused by the temperature being zero
    somewhere in the model. The relevant values for debugging are:
    temperature (0), adiabatic_temperature_gradient_for_viscosity (0) and 
    pressure (0).
```

anne-glerum commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -139,8 +141,29 @@ namespace aspect
+
         AssertThrow(this->get_parameters().enable_elasticity == true,
-                    ExcMessage ("Material model Viscoelastic only works if 'Enable elasticity' is set to true"));
+                    ExcMessage("Material model Viscoelastic only works if 'Enable elasticity' is set to true"));
```
elasticity.cc is already used for ViscoPlastic. 
rheology/visco_plastic.cc reads in a parameter `use_elasticity = prm.get_bool ("Include viscoelasticity");` and only when `use_elasticity` is true, `elastic_rheology` is initialized etc. I think that's why it works. I've found having to set `Enable elasticity` and `Include viscoelasticity` in two different places a bit unintuitive, but changing it would be a separate PR. We can discuss it with John.

anne-glerum commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -246,22 +275,44 @@ namespace aspect
-                // Get old stresses from compositional fields
-                SymmetricTensor<2,dim> stress_old;
+                // Get stress from timestep $t$ rotated and advected into the current
+                // timestep $t+\Delta te$ from the compositional fields.
```
Good catch!

anne-glerum commented on include/aspect/postprocess/stress_component_statistics.h:
```diff
@@ -0,0 +1,53 @@
+/*
```
It was part of this PR, but as discussed I will use it in a test and move it to tests/ or remove it completely.

anne-glerum commented on include/aspect/postprocess/visualization/stress_residual.h:
```diff
@@ -0,0 +1,65 @@
+/*
```
This pp is part of the PR.

gassmoeller commented:
@anne-glerum I tried to have a look at this PR, but you have a lot of unrelated changes in your branch (see https://github.com/geodynamics/aspect/pull/4370/files). It looks like you may have merged the 2.4.0 release branch, or some other changes. Can you rebase and try to remove the unrelated changes from your branch?

anne-glerum commented:
Hi @gassmoeller , yes I can see that makes reviewing very complicated. I rebased to the release because that is the version we wanted to use for the paper. 

I can make a copy of this branch (based on the release) and then rebase this PR to main. Would it be okay if I squash my commits before I rebase? I remember rebasing to the release being very laborious because of a multitude of commits giving merge conflicts.  

Alternatively, I update the copy of this branch I still have from before the rebase to the release with the changes I have made in the meantime and then rebase that to main. I'd then still prefer to squash the commits first. I'm not sure what is preferable or whether it makes a difference.

gassmoeller commented:
> I can make a copy of this branch (based on the release) and then rebase this PR to main. Would it be okay if I squash my commits before I rebase?

Yes, I think that is the best approach. Keep a copy based on the release (for reference in the paper), but rebase to main so that we can merge it. Squashing the commits is fine.

anne-glerum commented:
> > I can make a copy of this branch (based on the release) and then rebase this PR to main. Would it be okay if I squash my commits before I rebase?
> 
> Yes, I think that is the best approach. Keep a copy based on the release (for reference in the paper), but rebase to main so that we can merge it. Squashing the commits is fine.

Okay, thanks. I will do that and let you know when I'm done :)

anne-glerum commented:
Hi @gassmoeller, I have rebased to main. I also ran the third benchmark (2D square under simple shear for half of the model time) again; I summarize the results below. Maybe you could keep the issues listed below in mind when going through the code.
1. Results are independent of resolution for both fields and particles. 👍 
2. Results are independent of the number of nonlinear iterations allowed for fields. 👍 
3. The error during shearing decreases for both fields and particles with smaller timesteps. 👍 
4. The step in error when shearing is stopped decreasing for smaller timesteps for both fields and particles 👍 , but only up to a timestep size of 0.0025 s. 👎 
![3_viscoelastic_build-up_simple_shear_particles_dtc_isnot_dte_SS_dt](https://github.com/geodynamics/aspect/assets/7631624/8c52495f-8911-433d-8365-66712ac30ee2)
5. After shearing has stopped, the particle runs use all the allowed nonlinear iterations (NI). The more iterations are allowed, the larger the error. 👎 
![3_viscoelastic_build-up_simple_shear_particles_dtc_isnot_dte_SS_NI](https://github.com/geodynamics/aspect/assets/7631624/d44e2b2a-0983-4931-b366-93f9ba581ee1)
6. The error is not exactly the same between fields and particles 👎 , but the difference decreases with timestep size 👍 . After shearing has stopped, the difference is largely due to the step in error for particle runs.
![3_viscoelastic_build-up_simple_shear_fields_particles_dtc_isnot_dte_SS_test_dt_GR2](https://github.com/geodynamics/aspect/assets/7631624/c6ae5005-23b9-4d3c-a979-e11bd677c12d)

gassmoeller commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -291,76 +386,182 @@ namespace aspect
+                // Therefore we subtract the old stress stress_t.
+                for (unsigned int j = 0; j < SymmetricTensor<2, dim>::n_independent_components; ++j)
+                  {
+                    //out.reaction_terms[i][j] = -in.composition[i][j] + stress_0[SymmetricTensor<2, dim>::unrolled_to_component_indices(j)];
```
Did you leave this line in on purpose? If not: remove it.

gassmoeller commented on include/aspect/postprocess/stress_component_statistics.h:
```diff
@@ -0,0 +1,53 @@
+/*
+  Copyright (C) 2011 - 2022 by the authors of the ASPECT code.
```
My earlier comment about this file has not been addressed yet:
My question:
> Is this file part of this PR or was it added in another one that was merged into this PR?

Your reply:
> It was part of this PR, but as discussed I will use it in a test and move it to tests/ or remove it completely.

gassmoeller reviewed (commented):
I started a review, but didnt get through the important part yet, I will continue tomorrow or Wednesday. Just some comments for now.

gassmoeller commented:
Hi @anne-glerum, I think I understand the problem (after about 2 hours of reconstructing the order of operations :smile:) and have an idea for a fix, but it is a bit complex:
*Problem:*
By using the combination of particle properties `composition` and `elastic_stress`, and interpolating the resulting property back into a compositional field, we are circumventing the particle reset that happens at the start of every nonlinear iteration. Here is why: At the start of a NI, particle properties are reset. Then the particle property stress is overwritten from the solution vector in the `composition` property plugin. Then the `elastic_stress` plugin modifies that value. Then that value is interpolated back to the field into the solution vector. Now the cycle starts again, we reset particles, but then overwrite the properties with the solution vector (which was not reset). Therefore stress is infinitely accumulated in every iteration.
*Solution:*
I think this will be tricky, I dont see an easy solution. We cannot limit updating the particle properties from the field to the first nonlinear iteration, because in subsequent NI's the particles will be reset to their original values from the previous timestep again, so we would be missing out on the update that happened in the operator splitting step. I think we need to discuss three avenues and decide which one is most promising:
1. break the cycle of constant update somehow, e.g. by letting the `elastic_stress` particle property have its own set of properties (instead of reusing `composition`). I dont quite see if this will work, but it feels like it should be possible somehow.
2. Make operator splitting work within nonlinear iterations. Then we could move the operator splitting step into the loop, and presumably would reset the fields as well at the start of each NI (then we could use our current workflow, and updates would not accumulate over NIs).
3. Implement operator splitting on particles. We dont really need a full operator splitting, just a possibility to update particle properties before they are advected (e.g. via a signal). Then we could do the work we currently do in the operator splitting step directly in the particle property, and that would mean we dont need to update particle properties from the fields (and therefore the reset at the beginning of the NI would work correctly). This would mean elastic stress would be completely stored and computed on particles, and the compositional field would be a read-only convenience copy for the Stokes equation only.

Each solution is a bit tricky, but i think we need one of them. I prefer 2 or 3, because they also add additional functionality to ASPECT, and are not limited to our current problem. Maybe something we need to work on during the hackathon?

Could you try the problematic benchmark with `single Advection, iterated Stokes`? It may not be very accurate, but the solution for particles and fields should be much closer together?

anne-glerum commented:
Ohhh brilliant @gassmoeller! Thank you for taking the time to figure this out! 

Concerning option 3, we would need access to the `old_solution` at the location of particles to compute the operator splitting update, instead of the `solution` the plugins get now, but that's easily done.

It seems a bit like duplicating functionality (having an operator splitting for fields and for particles), but completely storing and computing stresses on the particles is a clean option. And, like you say, it's an update to a property value, it doesn't need the additional functionality of solving the ODEs over multiple timesteps. 

Maybe we can discuss it tomorrow. And yes, this does sound like a hack project. 



> Could you try the problematic benchmark with `single Advection, iterated Stokes`? It may not be very accurate, but the solution for particles and fields should be much closer together?

Using `iterated Advection and Stokes` with max one NI or `single Advection, iterated Stokes` gives the same result. It is indeed much closer to the fields, but still a bit off:
![3_viscoelastic_build-up_simple_shear_fields_particles_dtc_isnot_dte_SS_test_dt_GR2](https://github.com/geodynamics/aspect/assets/7631624/a3ebe63b-404f-4f45-b868-62614469c638)
(compare dark blue solid, dashed and dotted lines above)

![3_viscoelastic_build-up_simple_shear_particles_dtc_isnot_dte_SS_NI](https://github.com/geodynamics/aspect/assets/7631624/0287a6c1-0529-466c-8c10-ca05d42fb872)
(compare pink and blue line with vertical dash above)

anne-glerum commented:
We've been discussing the options you have suggested above in our meeting yesterday, and came up with some potential ways of implementing the third option. Here's how I see it could be done:
1. Add a function in helper_functions.cc (where the current operator splitting is housed). The function is called at the beginning of a nonlinear iteration, after the particles have been reset (e.g. after restore_particles() in line 1140 of solve_iterated_advection_and_stokes()). It is only called if `(particle_world.get() != nullptr)` and the `Use elasticity = true` (or we check for the two material models visco_elastic and visco_plastic, or there is a more general check for applying an update to the particle properties, or there is a signal) . The function loops over all particles, creating MaterialModelInput and MaterialModelOutput for each particle, filling the Input with the solutions from the last timestep at the particle's current (= old) position. Using these, the material model is evaluated. The returned reaction rate is multiplied with the current timestep and then added to the appropriate property (the stresses).
2. Instead of a function, we add a signal `post_restore_particles` that connects to the same location, so after particles have been restored. The functionality that calls the material model for the reaction_rates and updates the relevant particle properties is placed in the particle property plugin `elastic_stress`. Other particle properties could use the same signal.
3. We add in the update in the particle property `elastic_stress` before the reaction_terms are added. Downside is that the properties are updated *after* the particles have been advected, so we would have to retrieve the old particle locations (maybe from the particle world backup?) and evaluate the old solution at those locations.

Option 1 and 2 are the most straightforward I think. This would probably be a separate PR?

gassmoeller commented:
Option 2 sounds the most appealing to me, because it introduces the least amount of specific code into main ASPECT and almost all new code can go into the particle property plugin.

Long term we should unify the particle advection and particle update property to use the same integration scheme (currently we use one scheme for advection, and then do a single step update of the properties afterwards). But that is for another time.

anne-glerum commented:
> Option 2 sounds the most appealing to me, because it introduces the least amount of specific code into main ASPECT and almost all new code can go into the particle property plugin.

Alright! I can start with a separate PR that introduces the signal and has a test with a particle property plugin that uses the signal and does a simple update without calling the material model. Or shall I implement it directly in this PR so that we can see whether it fixes the issues?

anne-glerum commented:
Alright, I first implemented the signal and a new particle property that just uses the signal to increase its property by a constant every timestep. That worked, showing the correct particle property value after a certain amount of time. It's on this branch: https://github.com/anne-glerum/aspect/tree/signal_after_particle_reset

Then I used this signal in the current PR and modified the elastic_stresses particle property to 
1.  Initialize the stress tensor properties again (so the `composition` property is no longer used)
2. Update the particle stress values after the particles have been restored at the beginning of a nonlinear iteration
3. Use the stress values on the particles as material model input, not the field solutions, for both the new stress update using the material model reaction rates and the original property update that uses the material model reaction term

I also had to 
1. Change any AssertThrows that check `use_operator_splitting` to take into account that we can have it set to false, but still want to compute the reaction_rates in the material model. I used `(this->get_parameters().mapped_particle_properties).count(this->introspection().compositional_index_for_name("ve_stress_xx"))` as a condition to check whether the `elastic_stress` particle property needs the reaction rates, but I don't think that's the best way to do it, since the field name could be used without the particle property. Maybe we need an additional bool for the particles, or we check whether the specific particle plugin `elastic_stress` is used.
2. Use the stress tensor on the particles as the old stress tensor to compute the reaction term in the material model  

Although it can be improved, it does work!

The dotted line shows the new behaviour, which is very similar to only allowing one NI before:
![3_viscoelastic_build-up_simple_shear_fields_particles_dtc_isnot_dte_SS_test_dt_GR2_OS](https://github.com/geodynamics/aspect/assets/7631624/f76c09b0-da92-4dcb-af41-6913bdedb268)
Note that there is still a small difference between field and particle results.

The error no longer increases when increasing the max amount of NI (dashed lines all plot on top of each other):
![3_viscoelastic_build-up_simple_shear_particles_dtc_isnot_dte_SS_NI_noOS](https://github.com/geodynamics/aspect/assets/7631624/7657a2ba-701c-4c70-94ec-2229dc8cbe0b)

And smaller timesteps still decrease the error (except when dt=0.00125s after shearing has stopped):
![3_viscoelastic_build-up_simple_shear_particles_dtc_isnot_dte_SS_dt_noOS](https://github.com/geodynamics/aspect/assets/7631624/ef51241f-faec-43e7-b19b-9f40750901af)

anne-glerum commented:
Hi @gassmoeller and @bobmyhill , I've gone through and updated the documentation and addressed your comments. I'll fix the merge conflict tomorrow, but even before that I think this is ready for another look.

gassmoeller commented on source/postprocess/visualization/stress_residual.cc:
```diff
@@ -0,0 +1,166 @@
+        Assert(input_data.solution_values[0].size() == this->introspection().n_components, ExcInternalError());
+        Assert(input_data.solution_gradients[0].size() == this->introspection().n_components, ExcInternalError());
+
+        double dtc = this->get_timestep();
```
same comment, move below

gassmoeller commented on source/postprocess/visualization/stress_residual.cc:
```diff
@@ -0,0 +1,166 @@
+                                                  "for the difference between the second moment invariant "
+                                                  "of the deviatoric stress tensor and the yield stress. "
+                                                  "Note that this plugin currently only works when the "
+                                                  "'viscoplastic' material model is used. ")
```
```suggestion
                                                  "'visco plastic' material model is used. ")
```

gassmoeller commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -269,98 +371,293 @@ namespace aspect
 
+            // Get the current velocity gradients, which get
+            // updated in each nonlinear iteration.
+            // TODO Decide on whether to use W^t instead of W^(t+dtc)
+            // which is used now.
```
What about this todo?

anne-glerum replied:
I'll change it back to W^t, but I need to test that this is still the best option.

anne-glerum replied:
Still need to test this, sorry :)

gassmoeller reviewed (commented):
First set of comments, I still need to check particle/property/elastic_stress.cc and material_model/rheology/elasticity.cc
A few other points:
- we need a changelog entry

anne-glerum commented on source/material_model/visco_plastic.cc:
```diff
@@ -283,13 +280,62 @@ namespace aspect
+    ViscoPlastic<dim>::
+    get_elastic_viscosity(const double shear_modulus) const
+    {
+      if (rheology->use_elasticity)
```
It should be if this branch was rebased to main more recently :)

anne-glerum replied:
I'll rebase after I'm done with the shear heating.

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -531,6 +567,14 @@ namespace aspect
+                           Patterns::Bool(),
+                           "Whether to take the minimum of the diffusion and dislocation creep "
+                           "viscosity (true) or use a composite of both viscosities (false). "
+                           "When the full strain rate is used to compute each viscosity "
```
Since we don't properly distribute the strain rate between dislocation and diffusion creep (like it's done in diffusion_dislocation.cc), but use the full strain rate to compute both the dislocation and the diffusion viscosity, it's better to take the minimum of the two.

anne-glerum commented on source/material_model/viscoelastic.cc:
```diff
@@ -86,9 +93,32 @@ namespace aspect
+          ((this->get_parameters().mapped_particle_properties).count(this->introspection().compositional_index_for_name("ve_stress_xx"))))
+        elastic_rheology.fill_reaction_rates(in, average_elastic_shear_moduli, out);
+    }
 
```
The original cc file has only one line between the functions, so I stuck with that. I can update all spacing in the file.

anne-glerum commented on source/simulator/solver_schemes.cc:
```diff
@@ -992,6 +1033,10 @@ namespace aspect
         if ((particle_world.get() != nullptr) && (nonlinear_iteration > 0))
           particle_world->restore_particles();
 
+        // Apply a particle update if required by the particle properties
+        if (timestep_number > 0)
+          signals.post_restore_particles(*particle_world.get());
```
Yes, I'll move it into the if condition.

I didn't trigger the signal in timestep 0, because the operator splitting on fields doesn't do so either. And for the specific stress update that I want to do, there is no update to do of the previous timestep during timestep 0. I wrote this in the comment on the same call to the signal but in a different solver scheme. Anyway, I'll remove the requirement.

anne-glerum commented:
Hi Rene, I see this accumulated so merge conflicts, I'll update them tomorrow.

anne-glerum commented on source/particle/property/elastic_stress.cc:
```diff
@@ -22,6 +22,7 @@
 #include <aspect/material_model/visco_plastic.h>
 #include <aspect/material_model/viscoelastic.h>
 #include <aspect/initial_composition/interface.h>
+#include "../world.cc"
```
The function internal::construct_solution_evaluators is not declared in world.h, but only in world.cc. This was the dirty fix. Shall I add it to .h in a separate PR?

bobmyhill replied:
I'd say yes to this :)

anne-glerum replied:
As I replied to Rene, the function internal::construct_solution_evaluators is not declared in world.h, but only in world.cc, including the cc file was the dirty fix. I can add the function to the .h file in a separate PR?

bobmyhill replied:
Yes, I think that declaring the function in the header would be good.

bangerth replied:
You need to, or you will likely get linker errors on some systems. Including .cc files is not allowed.

anne-glerum commented:
Some test fail to run, I will address that asap.

bobmyhill commented:
@anne-glerum: if you're happy that the performance is good, I guess we can turn the elastic damper back on. Is the "different output" correct as far as you can tell?

anne-glerum commented:
Concerning the 36 failing tests, there are several reasons why they fail:

1. needs an elastic damper, but this is no longer allowed:
  229 - damped_viscoelastic_bending_beam 

2. Uses the DG limiter with a non-stress field. Therefore that field's advection equation doesn't convergence with the now multiple NI per timestep:
  285 - elastic_stabilization_time_scale_factor
  402 - infill_density_flexure
  1012 - viscoelastic_bending_beam
  1013 - viscoelastic_bending_beam_averaging 
  1014 - viscoelastic_bending_beam_gmg 
  1016 - viscoelastic_bending_beam_principal_stress
  1020 - viscoelastic_plate_flexure

3. Because of switching on DG, the number of DOFs has increased and therefore the output files (these don't need fixing):
  987 - visco_plastic_vep_stress_build-up_fixed_elastic_time_step
  1015 - viscoelastic_bending_beam_particles
  1017 - viscoelastic_fixed_elastic_time_step
  1018 - viscoelastic_numerical_elastic_time_step
  1019 - viscoelastic_numerical_elastic_time_step_particles
  1025 - viscoelastoplastic_yield_plastic_viscous_strain_weakening

4. Because of switching on an iterative solver scheme, now more (but small number of) NI are done (don't need fixing):
  1022 - viscoelastic_stress_averaging
  1023 - viscoelastic_stress_build-up
  1024 - viscoelastic_stress_build-up_gmg

TBC...

anne-glerum commented:
@gassmoeller What do you think about test 229 with the elastic damper that we have disabled in this PR? Could we change the test results to succeed if the Assert for the damper is triggered? Or is there a way to both preserve the current/main test output and have the test succeed with the new implementation?

For these tests:
319 - free_surface_timestep_repeat -> doesn't repeat
338 - gmg_mesh_deform_ghost_entries
812 - skip_refinement
I'll see what I can do to get them closer to the original results, but if that doesn't work out, how do you feel about removing the elasticity? Elasticity doesn't seem integral to what these tests are testing so I could remove it in a separate PR.

bobmyhill commented on source/material_model/visco_plastic.cc:
```diff
@@ -309,13 +309,56 @@ namespace aspect
+    {
+      if (this->get_parameters().enable_elasticity)
+        {
+          //TODO Check: Should this viscosity be scaled with the timestep ratio?
```
I guess the question is whether to modify the value here, or on each of the lines calling this function. I think you already made this decision in `material_model/rheology/visco_plastic.cc`?

Need to address this in `viscoelastic.cc` too

anne-glerum replied:
Yes, that's the question. calculate_elastic_viscosity in material_model/rheology/elasticity.cc does not scale with the timestep ratio, so in the rest of elasticity.cc it is done after calling calculate_elastic_viscosity. The function here, get_elastic_viscosity, is called by visualization postprocessors. It is less error-prone to do it in one place, which would be in calculate_elastic_viscosity in elasticity.cc. Would you agree?

bobmyhill replied:
Yes, I agree. I think that everywhere in the derivation the "elastic viscosity" term actually uses the computational timestep (i.e. Eq.51 in our draft), and that has the same right to being called the elastic viscosity as anything else.

anne-glerum replied:
Okay, done! I introduced a new function for the timestep ratio and scaled the elastic viscosity in the calculate_elastic_viscosity function, this cleaned up some duplicated code in several places.

bobmyhill commented on source/particle/property/elastic_stress.cc:
```diff
@@ -191,7 +471,12 @@ namespace aspect
-                                        "accumulated. See the viscoelastic material model "
+                                        "accumulated. This plugin modifies the properties "
+                                        "with the name ve_stress_*. It first applies the update "
+                                        "of the stresses of the previous computational timestep, "
```
This needs a bit more clarification. What does this first update do?

bobmyhill commented on source/postprocess/visualization/stress_residual.cc:
Could this visualization postprocessor be moved to another PR?

anne-glerum replied:
Yes, this postprocessor is new, so I can move it to another PR. The other two postprocessors already exist, I'm just adapting them to the new formulation, so I think they should stay in this PR.

anne-glerum replied:
Actually, that would mean first reverting the assembly of the stress tensor to match with main and then in a separate PR update it again to the formulation in this PR. Would you mind if this postprocessor also stays in this PR?

bobmyhill replied:
Ok, that's fine by me.

bobmyhill commented on source/postprocess/visualization/stress_second_invariant.cc:
Could this visualization postprocessor be moved to another PR?

bobmyhill commented on source/postprocess/visualization/surface_stress.cc:
Could this visualization postprocessor be moved to another PR?

anne-glerum commented on tests/visco_plastic_vep_plate_flexure.prm:
```diff
@@ -27,16 +27,16 @@ subsection Material model
     set Viscosity averaging scheme  = maximum composition
     set Cohesions                   = 1.e50
-    set Densities                   = 2700,  2700,  2700,  2700,  2700,  1890
+    set Densities                   = 2700,  2700,  2700,  2700,  2700,  2700,  2700,  2700,  1890
```
Yes, that would be great, see #5542.

anne-glerum commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -269,80 +355,199 @@ namespace aspect
+                  stress_0_t[SymmetricTensor<2, dim>::unrolled_to_component_indices(j)] = in.composition[i][j];
+
+                // Get the old stress that is used to interpolate to timestep $t+\Delta t_c$. It is stored on the
+                // second set of n_independent_components fields, e.g. in 2D on field 3, 4 and 5.
```
This is not entirely true. The old_stress in in.composition was advected but not rotated into the previous timestep. Below it will be updated to the full stress of the previous timestep, so that it can be advected into the current timestep (but again not rotated).

bobmyhill replied:
Ok good to know. Could you update the comment to say that?

anne-glerum commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -502,7 +534,9 @@ namespace aspect
           {
-            for (unsigned int i = 0; i < SymmetricTensor<2,dim>::n_independent_components ; ++i)
+            // First n_independent_components are the ve_stress_*, the next the ve_stress_*_old;
+            // they are the first fields, as is asserted in parse_parameters of elasticity.cc
```
Yes, I'll update after rebasing to main now that #5542 is merged.

anne-glerum commented on source/heating_model/shear_heating.cc:
```diff
@@ -76,6 +78,75 @@ namespace aspect
+
+              heating_model_outputs.heating_source_terms[q] = stress * visco_plastic_strain_rate;
+            }
+
```
It is not really duplicating calculations in the material model, but this assembly of the updated stress tensor is done in several plugins outside of the material model (here and in the visu postprocessors). I can imagine creating a function in the material model that returns the fully updated stress tensor of the current timestep, so that does lines 85-140 of shear_heating.cc.

bobmyhill replied:
Hmmm. I still think it might be better to put the bulk of this calculation in the material model, because:
- A lot of this work is already being done in `Elasticity<dim>::fill_elastic_force_outputs` and `Elasticity<dim>::fill_reaction_outputs`.
- A different formulation of the elasticity problem (a more material-oriented model, for example) might require a different formulation for the heating term.
- The elastic strain rate tensor is something we might want to use elsewhere (and/or visualise directly). For example, we might want to decompose the various components of the strain.

Could a viscous_dissipation term be computed in `elasticity.cc` and passed to an object in `ElasticOutputs` or `ElasticAdditionalOutputs`? Then `shear_heating.cc` could use that object:
`if (this->get_parameters().enable_elasticity == true)`
`   heating_model_outputs.heating_source_terms[q] = elastic_out->viscous_dissipation[i]`

I might be missing something that would make this hard to do.

anne-glerum commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -159,6 +178,16 @@ namespace aspect
+                                   "compositional field is called ve_stress_yy_old."));
+            AssertThrow(this->introspection().compositional_index_for_name("ve_stress_xy_old") == 5,
+                        ExcMessage("Rheology model Elasticity only works if the sixth "
+                                   "compositional field is called ve_stress_xy_old."));
```
yes, I'll update after rebasing to main now that #5542 is merged.

gassmoeller reviewed (commented):
A first batch of comments, I have to split this into several reviews because of the length of the PR.

gassmoeller reviewed (commented):
Next batch of comments mostly concerning the particle property and the postprocessors.

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -110,12 +110,31 @@ namespace aspect
+            // The first set of stresses in in.composition holds $\tau^{0adv}$
+            // after the first advection nonlinear iteration,
+            // which is when the viscosity matters for the Stokes system.
+            const std::vector<unsigned int> stress_field_indices = this->introspection().get_indices_for_fields_of_type(CompositionalFieldDescription::stress);
```
#5643 should allow you to improve this to:
```suggestion
            const std::vector<unsigned int> &stress_field_indices = this->introspection().get_indices_for_fields_of_type(CompositionalFieldDescription::stress);
```

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -110,12 +110,31 @@ namespace aspect
+            // after the first advection nonlinear iteration,
+            // which is when the viscosity matters for the Stokes system.
+            const std::vector<unsigned int> stress_field_indices = this->introspection().get_indices_for_fields_of_type(CompositionalFieldDescription::stress);
+            for (unsigned int j = 0; j < SymmetricTensor<2, dim>::n_independent_components; ++j)
+              stress_0_advected[SymmetricTensor<2, dim>::unrolled_to_component_indices(j)] = in.composition[i][stress_field_indices[j]];
+
+            // The old stresses are only changed in the operator splitting step and have been advected into
+            // the current timestep. They are stored in the second set of n_independent_components.
+            for (unsigned int j = 0; j < SymmetricTensor<2, dim>::n_independent_components; ++j)
+              stress_old[SymmetricTensor<2, dim>::unrolled_to_component_indices(j)] = in.composition[i][stress_field_indices[SymmetricTensor<2, dim>::n_independent_components+j]];
```
Again, depending on our convention of how to store the stress fields, this may be possible to simplify as in #5640.

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -208,8 +227,11 @@ namespace aspect
                   {
-                    non_yielding_viscosity = (viscosity_diffusion * viscosity_dislocation)/
-                                             (viscosity_diffusion + viscosity_dislocation);
+                    if (use_minimum_creep_viscosity)
```
I feel like it would be more consistent with the code structure to make this a new flow law separate from `composite`. Something like `composite_minimum_creep`.  That would save one additional input parameter and point out the difference between `composite` and `composite_minimum_creep`.

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -261,22 +286,21 @@ namespace aspect
-                    const SymmetricTensor<2,dim> effective_strain_rate =
-                      elastic_rheology.calculate_viscoelastic_strain_rate(in.strain_rate[i],
+
+                    // The second moment invariant is returned.
```
This comment at first confused me, nothing is "returned" and the process of computing the strain rate is explained in the comment in line 280. If you decide to keep this comment please move it one line down (where the second moment invariant is computed). Or you can delete it completely.

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -246,7 +268,10 @@ namespace aspect
               {
-                const std::vector<double> &elastic_shear_moduli = elastic_rheology.get_elastic_shear_moduli();
+                // Step 3a: calculate viscoelastic (effective) viscosity
+                // The preyield viscoelastic viscosity is scaled with the
+                // timestep ratio in the calculate_viscoelsatic_viscosity function.
```
Not sure this is the right place for this comment. It is an implementation details that would belong in the `calculate_viscoelastic_viscosity` function, in particular because the meaning of timestep ratio is not explained.

gassmoeller commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -501,8 +528,11 @@ namespace aspect
-              composition_mask.set(i,false);
+            // Set a mask for the 2*n_independent_components fields representing the viscoelastic
+            // stress tensor components.
+            const std::vector<unsigned int> stress_field_indices = this->introspection().get_indices_for_fields_of_type(CompositionalFieldDescription::stress);
+            for (auto it = stress_field_indices.begin(); it != stress_field_indices.end(); ++it)
+              composition_mask.set(*it, false);
           }
```
With #5643 you can simplify this code to:
```suggestion
            const std::vector<unsigned int> &stress_field_indices = this->introspection().get_indices_for_fields_of_type(CompositionalFieldDescription::stress);
            for (unsigned int field_index: stress_field_indices)
              composition_mask.set(field_index, false);
          }
```

But one additional question: The comment says to set a mask representing the viscoelastic stress tensor components, but to me it looks like the code is doing the opposite, it sets a mask that *excludes* all the stress tensor components. Do I misunderstand the code or does the comment need an update?

anne-glerum replied:
This whole function is actually no longer used since 
#5542 and should be removed.

gassmoeller commented on source/material_model/visco_plastic.cc:
```diff
@@ -303,8 +304,71 @@ namespace aspect
+          // Only fill them if operator splitting is used for fields or when the particles track
+          // the visco-elastic stresses.
+          if (this->get_parameters().use_operator_splitting ||
+              ((this->get_parameters().mapped_particle_properties).count(this->introspection().compositional_index_for_name("ve_stress_xx"))))
```
This condition could use an update to the compositional type machinery: `this->introspection().composition_type_exists(CompositionalFieldDescription::stress)`

anne-glerum replied:
I think the check can be removed. If elasticity is used with the VE or VP material model, we already do this check in parse_parameters in elasticity.cc. I did add the check `in.requests_property(MaterialProperties::reaction_rates))` in fill_reaction_rates.

gassmoeller commented on source/material_model/viscoelastic.cc:
```diff
@@ -72,26 +65,69 @@ namespace aspect
+      // timestep to the advected and rotated stress computed in the previous timestep ($\tau^{0adv}$)
+      // to obtain $\tau^{t}$.
+      if (this->get_parameters().use_operator_splitting ||
+          ((this->get_parameters().mapped_particle_properties).count(this->introspection().compositional_index_for_name("ve_stress_xx"))))
```
you can use the `introspection` to ask for material type here (see my comment in visco_plastic).

anne-glerum replied:
same here, I've removed this check.

anne-glerum commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -333,104 +451,218 @@ namespace aspect
+                {
+                  for (unsigned int c = 0; c < n_independent_components; ++c)
+                    stress_t[i][SymmetricTensor<2, dim>::unrolled_to_component_indices(c)] = in.composition[i][stress_field_indices[c]];
+                }
```
Yes and yes. Would it also be good to put the velocity evaluator (before line 487) in a separate function?

anne-glerum commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -459,6 +694,21 @@ namespace aspect
+        double dtc = this->get_timestep();
+        if (!this->simulator_is_past_initialization() ||
+            (this->get_timestep_number() == 0 && this->get_timestep() == 0))
+          dtc = std::min(std::min(this->get_parameters().maximum_time_step, this->get_parameters().maximum_first_time_step), elastic_timestep());
```
It's a bit complicated, but in all our benchmarks we set the maximum timestep to either equal the elastic timestep or to be a specific factor smaller than the elastic timestep. If the initial stress is nonzero, the timestep ratio matters for the force term and viscous dissipaton, and the timestep ratio also feeds into the elastic viscosity.

anne-glerum commented on source/particle/property/elastic_stress.cc:
```diff
@@ -54,6 +51,240 @@ namespace aspect
+
+                          // For the ve_stress_* fields, we use the values on the particles. After they have been restored,
+                          // their properties have the values of the previous timestep.
+                          for (unsigned int n = 0; n < this->n_compositional_fields(); ++n)
+                            material_inputs.composition[0][n] = old_solution[this->introspection().component_indices.compositional_fields[n]];
+                          for (unsigned int n = 0; n < 2*SymmetricTensor<2,dim>::n_independent_components; ++n)
+                            material_inputs.composition[0][n] = particle->get_properties()[data_position + n];
```
Yes, I've started to move away from the stress fields having to be the first fields. I'll make sure that is consistently done everywhere.

I now only fill the non-stress fields with the old_solution, so there is no overwriting with the particle values anymore.

anne-glerum commented on source/postprocess/visualization/principal_stress.cc:
```diff
@@ -107,44 +110,82 @@ namespace aspect
+
+                stress_old[0][0] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xx_old")];
+                stress_old[1][1] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_yy_old")];
+                stress_old[0][1] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xy_old")];
```
The implementation in main takes the stress directly from the stored fields and multiplies it with -1 to get the right sign convention. This branch's implementation builds the full shear stress from the fields stress_0 and stress_old (not negating them) and then multiplies the whole shear stress with -1. So the signs are the same?

gassmoeller reviewed (commented):
I didnt make it through all files, but here are some smaller suggestions. I think this is getting quite close to merge. Let's talk a bit more about the field duplication.

anne-glerum commented on include/aspect/simulator.h:
```diff
@@ -2097,6 +2097,7 @@ namespace aspect
       LinearAlgebra::BlockVector                                solution;
       LinearAlgebra::BlockVector                                old_solution;
       LinearAlgebra::BlockVector                                old_old_solution;
+      LinearAlgebra::BlockVector                                unlimited_solution;
```
I've removed everything about the unlimited_solution now that limited fields are excluded from the residual.

anne-glerum commented on include/aspect/material_model/rheology/elasticity.h:
```diff
@@ -52,6 +52,28 @@ namespace aspect
+        * at the evaluation points passed to the instance of
+        * MaterialModel::Interface::evaluate() that fills the current object.
+        */
+        std::vector<double> timestep_ratio;
```
Yes, a double would be sufficient. We could also remove it alltogether, as now that the`deviatoric_stress` is added to the ElasticAdditionalOutputs, there are no places that need the timestep_ratio.

anne-glerum commented on source/simulator/solver_schemes.cc:
```diff
@@ -337,6 +337,60 @@ namespace aspect
+    // for the fields that belong to that tensor. In other words, we compute an
+    // averaged initial residual using those fields that belong to the ve_stress tensor,
+    // and one for those belonging to the ve_stress_old tensor.
+    // TODO ve_stress_old values are not independent components of the solution vector,
+    // do we need to compute a residual for them at all?
```
Okay, I'll now have a look whether we can get around the ve_stress_old fields by applying the rotation to the advected fields whenever we need them, otherwise I can set their residuals to 0.

anne-glerum replied:
I've set their residuals to 0, as we decided to keep the ve_stress_old fields, so that we can extract the velocity gradients at their correct locations.

bobmyhill commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -474,7 +773,17 @@ namespace aspect
+        // use that instead.
+        // TODO: what to do with the damper? It is asserted to be zero when using
+        // this rheology in the visco_plastic and viscoelastic material models. But should
+        // the scaling apply to the dampened viscosity or the undampened viscosity?
```
I think this is a question about which value the user should input, the damper viscosity over the computational timestep `elastic_damper_viscosity_c` or over the elastic timestep `elastic_damper_viscosity_e`. I think it makes most sense for the user to input `elastic_damper_viscosity_c` over the computational timestep. This will be more "physical", not produce odd behaviour if the elastic timestep changes, and simulate a "Voigt" element with an isoviscous damper). 

Over the elastic timestep, we have `eta_el_e = shear_modulus*elastic_timestep() + elastic_damper_viscosity_e`
Also, from the derivation in our paper:
<img width="96" alt="image" src="https://github.com/geodynamics/aspect/assets/8144905/b5835c7b-134c-4a0d-bb5b-c93adbc8294a">

<img width="86" alt="image" src="https://github.com/geodynamics/aspect/assets/8144905/32b6fad4-22a0-4079-92db-16474d8db1f8">

So this implies that:
`shear_modulus*elastic_timestep()*timestep_ratio + elastic_damper_viscosity`
which is what you have in the code already.

anne-glerum replied:
Great, thanks!

bobmyhill commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -485,8 +794,14 @@ namespace aspect
+        // The viscous viscosity has not been scaled yet, so we do it here.
+        // Scaling both viscosities with the timestep ratio before computing the effective
+        // viscoelastic viscosity equals scaling the viscoelastic viscosity computed from
+        // the unscaled elastic and viscous viscosity.
```
Perhaps this is a good place to explain how this works. From our paper:
<img width="598" alt="image" src="https://github.com/geodynamics/aspect/assets/8144905/f8988d96-8c11-4f42-b7b8-a11f62011ecf">

bobmyhill commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -110,12 +110,32 @@ namespace aspect
+        SymmetricTensor<2, dim> stress_0_advected = numbers::signaling_nan<SymmetricTensor<2, dim>>();
+        SymmetricTensor<2, dim> stress_old = numbers::signaling_nan<SymmetricTensor<2, dim>>();
+        double elastic_shear_modulus = numbers::signaling_nan<double>();
         if (this->get_parameters().enable_elasticity)
```
Can all of the code in this if statement be moved into elasticity.cc?

anne-glerum replied:
It could be, but the stress tensors are used in a later if statement to compute the effective strain rate. That function call could get the `MaterialInputs in` and unwrap the tensors from the fields itself. It would do it for each volume fraction though.

bobmyhill commented on source/material_model/rheology/visco_plastic.cc:
```diff
@@ -110,12 +110,32 @@ namespace aspect
+            // (different rotations, different stress changes), but this is inconsistent with storing only one stress tensor.
+            // The averaging used is the same as for the viscosity.
+            const std::vector<double> &elastic_shear_moduli = elastic_rheology.get_elastic_shear_moduli();
+            elastic_shear_modulus = MaterialUtilities::average_value(volume_fractions, elastic_shear_moduli, viscosity_averaging);
```
Might we want to have a separate `shear_modulus_averaging` parameter declared in the elastic rheology module?

bobmyhill reviewed (changes requested):
Awesome work. A few very minor suggestions, but this looks almost good to go.

anne-glerum commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -150,61 +176,114 @@ namespace aspect
 
-        AssertThrow(this->get_parameters().enable_elasticity == true,
-                    ExcMessage ("Material model Viscoelastic only works if 'Enable elasticity' is set to true"));
+        if (Plugins::plugin_type_matches<MaterialModel::ViscoPlastic<dim>>(this->get_material_model()) ||
```
Nope, you're right, the if statement around these asserts can go.

gassmoeller commented:
@anne-glerum let us know when you want us to take another look at this.

gassmoeller commented:
Hi Anne,

> I've added material averaging to the viscosity in the reaction rates (see https://community.geodynamics.org/t/amg-vs-gmg-solver-inconsistency-in-3d-elasticity-models-in-spherical-domains/4017/19)

Ok.

> I've rebased to main and due to changes there had to add some requests_property logic in the material models.

Yes thats a result of another PR.
    
> At the moment, reaction_rates for elasticity are created when create_additional_named_outputs() is called (same for melt_global). This means that at the moment, reaction_rates for elasticity are created when create_additional_named_outputs() is called (same for melt_global). This means that the visualization postprocessor named_additional_outputs, for example, outputs reaction rates. At time zero, reaction_rates are not computed though, so the output will be nans then. See eg tests vep_chemical_compositions and viscoelastoplastic_yield_plastic_viscous_strain_weakening.

I think they should just be 0 in Timestep 0, that should be a simple change at the place where they are computed.

> Should MaterialProperties::additional_outputs include reaction_terms and reaction_rates? They have there own MaterialProperties enum and are added on top of additional_outputs to create all_properties in material_model/interface.h. But that means that asking for additional_outputs only in eg the visualization postprocessor named_additional_outputs and checking for the reaction_rates property only when filling the reaction rates, leaves them nan.

I would prefer if reaction_terms and reaction_rates are kept separate from additional outputs. That reaction_rates are of a type of additional outputs is more an accident.

> When material averaging is used + particles to track the stresses, how can we apply the same material averaged viscosity for the reaction rates as for the viscosity in the Stokes equation? The particle property plugin only calls the material model for one particle at the time, so the material model can't average over the cell. However, we do know all the particles in the cell, so maybe we can use them to somehow average?

Take a look at how the grain size particle property computes material outputs for all particles of a cell at the same time, maybe that works for the elastic stress property as well?

> I've gone through the test results and all tests run except one. This is a test with an elastic damper, which we no longer allow. Some tests show a significant change (e.g. factor 10) in RMS and max velocity in the initial timestep. With each timestep the difference becomes smaller. Often, the velocities are already big in main, and of course we changed the implementation, so I don't know if this difference is something I should investigate further.

Since our benchmarks show our new implementation to be better and different I wouldnt be to concerned about the test changes. As long as nothing explodes and the new results is not completely unreasonable it should be fine.

gassmoeller commented on include/aspect/particle/property/elastic_stress.h:
```diff
@@ -91,9 +113,35 @@ namespace aspect
+
+          /**
+           * The weight given to the stress values stored on the particles in the
+           * weigthed average with the stress values interpolated from the compositional
```
```suggestion
           * weighted average with the stress values interpolated from the compositional
```

gassmoeller commented on source/material_model/viscoelastic.cc:
```diff
@@ -73,27 +66,43 @@ namespace aspect
-              {
-                elastic_out->elastic_shear_moduli[i] = average_elastic_shear_moduli[i];
-              }
+          // If we have multiple compositions, we need to first compute their respective their viscoelastic viscosities,
```
```suggestion
          // If we have multiple compositions, we need to first compute their respective viscoelastic viscosities,
```

gassmoeller commented on source/particle/property/elastic_stress.cc:
```diff
@@ -54,6 +53,205 @@ namespace aspect
+        // Get the indices of all compositions that do not correspond to stress tensor elements.
+        std::vector<unsigned int> all_field_indices(this->n_compositional_fields());
+        std::iota (std::begin(all_field_indices), std::end(all_field_indices), 0);
+        set_difference(all_field_indices.begin(), all_field_indices.end(),
```
```suggestion
        std::set_difference(all_field_indices.begin(), all_field_indices.end(),
```

anne-glerum commented:
Hi @gassmoeller, I've addressed your comments and looked at the test results:
1. The damper test fails because we do not allow using an elastic damper anymore. I've marked the test as expected to fail.
2. The test `free_surface_timestep_repeat` is supposed to repeat timestep 1 using the `repeat_on_cutback` scheme, but with the PR implementation, the RMS/max velocity now decreases monotonically and there is therefore no need to repeat. I tried to trigger a repeat by increasing the timestep and changing the CFL condition, but didn't manage. The stored RMS looks like this, which increases again at t5 even though the loading is constant for this time period.
```
*** Timestep 0:  t=0 years, dt=0 years
*** Timestep 0:  t=0 years, dt=0 years
     RMS, max velocity:         4.3 m/year, 13.6 m/year
*** Timestep 1:  t=0.921264 years, dt=0.921264 years
*** Timestep 1:  t=0.460632 years, dt=0.460632 years
     RMS, max velocity:         19.5 m/year, 63.2 m/year
*** Timestep 2:  t=0.960632 years, dt=0.5 years
     RMS, max velocity:         8.63 m/year, 27.4 m/year
*** Timestep 3:  t=1.46063 years, dt=0.5 years
     RMS, max velocity:         7.63 m/year, 24.4 m/year
*** Timestep 4:  t=1.97242 years, dt=0.511791 years
     RMS, max velocity:         3.72 m/year, 11.6 m/year
*** Timestep 5:  t=2 years, dt=0.0275774 years
     RMS, max velocity:         65.2 m/year, 203 m/year
```
3. The test `viscoelastic_bending_beam_particles` doesn't converge in timestep 1, but does in timestep 2. As the resolution is reduced for the test, the resolution is the same as the beam thickness. Instead of trying to make it converge, I've limited the Max nr of nonlinear iterations to 5 instead of 100 to not waste time. 5 iterations more than the other timesteps need that do converge.
4. The tests `gmg_mesh_deform_ghost_entries` (traction load + fixed elastic timestep), `infill_density_flexure` (free surface + traction load + fixed elastic timestep) and `skip_refinement` (free surface + traction load) all show RMS/max velocities that differ up to a factor 20 with the old test results. The difference in velocity decays quickly though; I don't think we need to change the setup.

What do you want me to do about point 2?

gassmoeller commented:
Thanks! Concerning 2. the test is supposed to show that cutback works in models with mesh deformation, not necessarily that `repeat on cutback` works. Can you instead limit the number of nonlinear iterations to 2 (or something small) and force cutback by `repeat on nonlinear solver failure`? (`repeat on cutback` is still tested elsewhere)

anne-glerum commented:
Good point. I've adapted the test to repeat on nonlinear solver fail. I've also updated the test results and now the tester is happy :)

anne-glerum commented:
Thank you Rene!
